src/GafferScene/EvaluateLightLinks.cpp

```diff
@@ -0,0 +1,127 @@
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include <cstdio>
+
+#include "IECore/CompoundData.h"
+
+#include "GafferScene/EvaluateLightLinks.h"
+#include "GafferScene/SetAlgo.h"
+
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferScene;
+
+IE_CORE_DEFINERUNTIMETYPED( EvaluateLightLinks );
+
+EvaluateLightLinks::EvaluateLightLinks( const std::string &name )
+	:	SceneProcessor( name )
+{
+	m_lightLinkAttrName = "gaffer:linkedLights";
+
+	// Fast pass-throughs for the things we don't alter.
+	outPlug()->boundPlug()->setInput( inPlug()->boundPlug() );
+	outPlug()->transformPlug()->setInput( inPlug()->transformPlug() );
+	outPlug()->setNamesPlug()->setInput( inPlug()->setNamesPlug() );
+	outPlug()->setPlug()->setInput( inPlug()->setPlug() );
+	outPlug()->objectPlug()->setInput( inPlug()->objectPlug() );
+	outPlug()->childNamesPlug()->setInput( inPlug()->childNamesPlug() );
+	outPlug()->globalsPlug()->setInput( inPlug()->globalsPlug() );
+}
+
+EvaluateLightLinks::~EvaluateLightLinks()
+{
+}
+
+void EvaluateLightLinks::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneProcessor::affects( input, outputs );
+
+	if( input == inPlug()->attributesPlug() || input == inPlug()->setPlug() || input == inPlug()->globalsPlug() )
+	{
+		outputs.push_back( outPlug()->attributesPlug() );
+	}
+}
+
+void EvaluateLightLinks::hashAttributes( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
+{
+	SceneProcessor::hashAttributes( path, context, parent, h );
+
+	inPlug()->attributesPlug()->hash( h );
+
+	IECore::ConstCompoundObjectPtr attributes = inPlug()->attributesPlug()->getValue();
+
+	CompoundObject::ObjectMap::const_iterator it = attributes->members().find( m_lightLinkAttrName );
+	if( it != attributes->members().end() )
+	{
+		if( StringDataPtr lightsExpressionData = IECore::runTimeCast<StringData>( it->second ) )
+		{
+			SetAlgo::hashSetExpression( lightsExpressionData->readable(), inPlug(), h );
+		}
+	}
+  // IECore::ConstInternedStringVectorDataPtr setNames = inPlug()->setNamesPlug()->getValue();
+	// const std::vector<IECore::InternedString> setNamesReadable = setNames->readable();
+	// for( std::vector<IECore::InternedString>::size_type i = 0; i < setNamesReadable.size(); ++i )
+	// {
+	// 	h.append( parent->setHash( setNamesReadable[i] ) );
+	// 	// SceneProcessor::hashSet( setNamesReadable[i], context, parent, h );
+	// }
+}
+
+IECore::ConstCompoundObjectPtr EvaluateLightLinks::computeAttributes( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent ) const
+{
+	IECore::ConstCompoundObjectPtr attributes = inPlug()->attributesPlug()->getValue();
+
+	CompoundObject::ObjectMap::const_iterator it = attributes->members().find( m_lightLinkAttrName );
+	if( it != attributes->members().end() )
```

Comment:
I'm a big fan of returning early from functions rather than nesting deeper and deeper into conditionals like this - it makes the code much easier to follow, because you're not having to maintain a mental stack of "reasons I'm in this code path". Mirroring the hash example code above would also make it much easier to verify they match :

```
ConstCompoundObjectPtr inputAttributes = inPlug()->attributesPlug()->getValue( & inputHash );
ConstStringDataPtr expressionData = attributes->member<StringData>( "linkedLights" );
if( !expressionData )
{
    // Pass through
    return inputAttributes;
}

// Real work here, no nesting.
```